api/src/main/java/org/openmrs/api/impl/PersonServiceImpl.java

```diff
@@ -991,4 +995,272 @@ public void checkIfPersonAttributeTypesAreLocked() {
+			if (!attr.getVoided()) {
+				PersonAttribute tmpAttr = attr.copy();
+				tmpAttr.setPerson(null);
+				tmpAttr.setUuid(UUID.randomUUID().toString());
+				preferred.addAttribute(tmpAttr);
+				mergedData.addCreatedAttribute(tmpAttr.getUuid());
+			}
+		}
+	}
+	
+	private void mergeGenderInformation(Person preferred, Person notPreferred, PersonMergeLogData mergedData) {
+		// move all gender info
+		mergedData.setPriorGender(preferred.getGender());
+		if (!"M".equals(preferred.getGender()) && !"F".equals(preferred.getGender())) {
+			preferred.setGender(notPreferred.getGender());
+		}
+	}
+	
+	private void mergeNames(Person preferred, Person notPreferred, PersonMergeLogData mergedData) {
+		// move all names
+		// (must be done after all calls to services above so hbm doesn't try to save things prematurely (hacky)
+		for (PersonName newName : notPreferred.getNames()) {
+			boolean containsName = false;
+			for (PersonName currentName : preferred.getNames()) {
+				containsName = currentName.equalsContent(newName);
+				if (containsName) {
+					break;
+				}
+			}
+			if (!containsName) {
+				PersonName tmpName = constructTemporaryName(newName);
+				preferred.addName(tmpName);
+				mergedData.addCreatedName(tmpName.getUuid());
+				log.debug("Merging name " + newName.getGivenName() + " to " + preferred.getPersonId());
+			}
+		}
+	}
+	
+	private PersonName constructTemporaryName(PersonName newName) {
+		PersonName tmpName = PersonName.newInstance(newName);
+		tmpName.setPersonNameId(null);
+		tmpName.setVoided(false);
+		tmpName.setVoidedBy(null);
+		tmpName.setVoidReason(null);
+		// we don't want to change the preferred name of the preferred patient
+		tmpName.setPreferred(false);
+		tmpName.setUuid(UUID.randomUUID().toString());
+		return tmpName;
+	}
+	
+	private void mergeAddresses(Person preferred, Person notPreferred, PersonMergeLogData mergedData)
+	        throws SerializationException {
+		// move all addresses
+		// (must be done after all calls to services above so hbm doesn't try to save things prematurely (hacky)
+		for (PersonAddress newAddress : notPreferred.getAddresses()) {
+			boolean containsAddress = false;
+			for (PersonAddress currentAddress : preferred.getAddresses()) {
+				containsAddress = currentAddress.equalsContent(newAddress);
+				if (containsAddress) {
+					break;
+				}
+			}
+			if (!containsAddress) {
+				PersonAddress tmpAddress = (PersonAddress) newAddress.clone();
+				tmpAddress.setPersonAddressId(null);
+				tmpAddress.setVoided(false);
+				tmpAddress.setVoidedBy(null);
+				tmpAddress.setVoidReason(null);
+				tmpAddress.setPreferred(false); // addresses from non-preferred patient shouldn't be marked as preferred
+				tmpAddress.setUuid(UUID.randomUUID().toString());
+				preferred.addAddress(tmpAddress);
+				mergedData.addCreatedAddress(tmpAddress.getUuid());
+				log.debug("Merging address " + newAddress.getPersonAddressId() + " to " + preferred.getPersonId());
+			}
+		}
+		
```

Comment:
This method could be break in to several methods.  name of the method is called mergeAddresses but there are lot of things happening apart from merging addresses.